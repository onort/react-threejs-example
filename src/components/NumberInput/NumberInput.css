.container {
  margin: 0;
  display: flex;
}

/* Removes number input spinners */
input[type="number"]::-webkit-inner-spin-button,
input[type="number"]::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

/* Removes number input spinners */
input[type="number"] {
  -moz-appearance: textfield;
}

input:focus {
  outline: 0;
}

.numberInput {
  width: 2rem;
  line-height: 1.5;
  margin: 0;
  padding: 0.25rem;
  border: 1px solid var(--light-border);
}

.counter {
  display: inline-block;
  opacity: 0.66;
  transition: all 0.4s ease-in-out;
}

.counter:hover {
  opacity: 1;
}

.increase,
.decrease {
  display: block;
  padding: 0.1rem 0.2rem;
  line-height: 0.8;
  font-size: var(--text-lg);
  text-align: center;
  border: 1px solid var(--light-border);
  border-left: none;
  cursor: pointer;
  user-select: none;
}

.decrease {
  border-top: none;
}

.rangeLimits {
  align-self: center;
  margin-left: var(--spacer-sm);
  font-size: var(--text-sm);
  color: var(--muted-text);
}
