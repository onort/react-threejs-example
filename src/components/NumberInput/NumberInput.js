import React from "react"
import PropTypes from "prop-types"

import styles from "./NumberInput.css"

const NumberInput = props => {
  const { currentValue, max, min, onChange, onDecrease, onIncrease } = props
  const numberValue = currentValue !== null ? currentValue : ""
  return (
    <div
      className={`${styles.container} ${
        props.className ? props.className : ""
      }`}
    >
      <input
        className={styles.numberInput}
        onChange={onChange}
        type="number"
        value={numberValue}
      />
      <div className={styles.counter}>
        <span className={styles.increase} onClick={onIncrease}>
          &#43;
        </span>
        <span className={styles.decrease} onClick={onDecrease}>
          &#8722;
        </span>
      </div>
      {max && min && (
        <span className={styles.rangeLimits}>
          ({min} - {max})
        </span>
      )}
    </div>
  )
}

NumberInput.propTypes = {
  className: PropTypes.string,
  currentValue: PropTypes.number,
  max: PropTypes.number,
  min: PropTypes.number,
  onChange: PropTypes.func.isRequired,
  onDecrease: PropTypes.func.isRequired,
  onIncrease: PropTypes.func.isRequired
}

NumberInput.defaultProps = {
  currentValue: 0
}

export default NumberInput
