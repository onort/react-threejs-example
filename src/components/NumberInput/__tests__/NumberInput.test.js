import React from "react"
import { shallow } from "enzyme"

import NumberInput from "../"

describe("<NumberInput />", () => {
  it("should match snapshot", () => {
    const fn = jest.fn()
    const wrapper = shallow(
      <NumberInput
        currentValue={1}
        max={5}
        min={1}
        onChange={fn}
        onDecrease={fn}
        onIncrease={fn}
      />
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should fire onChange event if input is changed", () => {
    const fn = jest.fn()
    const onChange = jest.fn()
    const changeEvent = { target: { value: 2 } }
    const wrapper = shallow(
      <NumberInput
        currentValue={1}
        max={5}
        min={1}
        onChange={onChange}
        onDecrease={fn}
        onIncrease={fn}
      />
    )
    wrapper.find("input").simulate("change", changeEvent)
    expect(onChange).toBeCalledTimes(1)
    expect(onChange).toBeCalledWith(changeEvent)
  })

  it("should fire onDecrease event if decrease button is clicked", () => {
    const fn = jest.fn()
    const onDecrease = jest.fn()
    const wrapper = shallow(
      <NumberInput
        currentValue={1}
        max={5}
        min={1}
        onChange={fn}
        onDecrease={onDecrease}
        onIncrease={fn}
      />
    )
    wrapper.find(".decrease").simulate("click")
    wrapper.find(".decrease").simulate("click")
    wrapper.find(".decrease").simulate("click")
    expect(onDecrease).toBeCalledTimes(3)
  })

  it("should fire onIncrease event if increase button is clicked", () => {
    const fn = jest.fn()
    const onIncrease = jest.fn()
    const wrapper = shallow(
      <NumberInput
        currentValue={1}
        max={5}
        min={1}
        onChange={fn}
        onDecrease={fn}
        onIncrease={onIncrease}
      />
    )
    wrapper.find(".increase").simulate("click")
    wrapper.find(".increase").simulate("click")
    expect(onIncrease).toBeCalledTimes(2)
  })

  it("should render min-max range if provided", () => {
    const fn = jest.fn()
    const max = 10
    const min = 0
    const wrapper = shallow(
      <NumberInput
        currentValue={1}
        max={max}
        min={min}
        onChange={fn}
        onDecrease={fn}
        onIncrease={fn}
      />
    )
    expect(wrapper.find(".rangeLimits").exists()).toBe(true)
    expect(wrapper.find(".rangeLimits").text()).toBe(`(${min} - ${max})`)
  })

  it("should have a custom className", () => {
    const testClassName = "testClassName"
    const fn = jest.fn()
    const wrapper = shallow(
      <NumberInput
        className={testClassName}
        currentValue={1}
        max={5}
        min={1}
        onChange={fn}
        onDecrease={fn}
        onIncrease={fn}
      />
    )
    expect(wrapper.find(`.${testClassName}`).exists()).toBe(true)
  })
})
