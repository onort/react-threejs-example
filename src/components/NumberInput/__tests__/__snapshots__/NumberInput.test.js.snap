// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberInput /> should match snapshot 1`] = `
<div
  className="container "
>
  <input
    className="numberInput"
    onChange={[MockFunction]}
    type="number"
    value={1}
  />
  <div
    className="counter"
  >
    <span
      className="increase"
      onClick={[MockFunction]}
    >
      +
    </span>
    <span
      className="decrease"
      onClick={[MockFunction]}
    >
      −
    </span>
  </div>
  <span
    className="rangeLimits"
  >
    (
    1
     - 
    5
    )
  </span>
</div>
`;

exports[`<NumberInput /> should match snapshot 2`] = `"<div class=\\"container \\"><input type=\\"number\\" class=\\"numberInput\\" value=\\"1\\"/><div class=\\"counter\\"><span class=\\"increase\\">+</span><span class=\\"decrease\\">−</span></div><span class=\\"rangeLimits\\">(1 - 5)</span></div>"`;
